### PR TITLE
Enhance views service 

### PR DESCRIPTION
**Pull request by sbatten**
refs https://github.com/microsoft/vscode/issues/87461

@sandy081 let me know if this looks on the right path. I ran into an issue with the views service now depending on the panel service and the panel service depending on the views service, both in the constructor. you can see I work around this with the instantiation service but still the panel service needs delay its panel registration call somehow (see setTimeout call in constructor). I don't think this part is quite right so I'm looking for suggestions. Depending on a few things, sometimes I see the outline view rendered twice in the panel when moved during testing so need to iron that out as well. I think it might just be caused by the setTimeout though.

### CODE REVIEW

**Comment by sandy081** [comment]
Went through the changes. They make sense and I have following suggestions:

1. Now, `ViewsService` and `ViewsRegistry` are having similar functionality - registering views and providing events. Is it possible to extract that into a model say `ViewDescriptorsModel` that can store view descriptors and provide events when something happens? Then. ViewsService and ViewsRegistry can simply use this model as composition and duplication of code can 

2. To your issues regarding cyclic dependency between `ViewsService` and `PanelService` - I would suggest to separate `PanelPart` and `PanelService` just like `ActivityBarPart` and `SidebarPart`.

**Review by sandy081 — CHANGES_REQUESTED** [review]
Provided feedback in the comment

**Comment by isidorn** [comment]
@sbatten If the splitup of the `PanelPart` makes your life easier than I am fully suporting this. For lack of a better name you would introduce something like `PanelTitlePart`. The downside of this approach is that from the UX perspective they do not match, since the ActivityPart and the SidebarPart are really different and you can have one without the other, we do not have a notion like that in the Panel. 
Bottom line: we should get rid of the timeout, if there are no better option than splitting up works for me.

**Comment by sbatten** [comment]
@sandy081 @isidorn I explored splitting up the panelpart and panelservice as well as alternatives. I settled on delaying the viewsservice's usage of panel service by using the panel registry directly. its not the perfect solution but i found it better than the timeout as well as significantly simpler than the splitting of the panelpart/service.

**Comment by sbatten** [comment]
@sandy081 I'm not sure I understand the model suggestion above. The interfaces look similar but the actual logic is a bit different. When I tried to logic it out, I felt like I was adding an equivalent amount of code with not much benefit. Perhaps we should have a call in the morning.

File: src/vs/workbench/browser/parts/panel/panelPart.ts [hunk 1]
```diff
@@ -281,6 +281,7 @@ export class PanelPart extends CompositePart<Panel> implements IPanelService {
 	}
 
 	private onPanelOpen(panel: IPanel): void {
+		const viewsService = this.instantiationService.invokeFunction(accessor => accessor.get(IViewsService));
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/panel/panelPart.ts]: This can be removed I suppose as you are having `IViewService` defined in the constructor

File: src/vs/workbench/browser/parts/views/views.ts [hunk 2]
```diff
@@ -678,23 +699,64 @@ export class ViewsService extends Disposable implements IViewsService {
...
+			this._onDidChangeContainer.fire({ views, from, to });
+		}
+	}
+
+	getViews(container: ViewContainer): IViewDescriptor[] {
+		return this.viewsRegistry.getViews(container);
+	}
+
+	private async openComposite(compositeId: string, location: ViewContainerLocation, focus?: boolean): Promise<IPaneComposite | undefined> {
+		const panelService = this.instantiationService.invokeFunction(accessor => accessor.get(IPanelService));
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/views.ts]: I do not think this is correct usage of services. If it is difficult to extract panel part, how about extracting views service into `IViewDescriptorService` and `IViewsService` ? Former deals only with view and view container descriptors and later deals with views (opening, focusing etc) ?

**Review by sandy081 — CHANGES_REQUESTED** [review]

**Comment by sandy081** [comment]
Re model suggestions:

ViewsService and ViewsRegistry are maintaining stores for view descriptors. My suggestion is to have a `ViewDescriptorStore` that can store view descriptors per container and trigger events when there are changes so that the registry and service can use this store?

Not sure if I will be available for call today as I might be leaving early.

**Review by sandy081 — APPROVED** [review]
LGTM. I would prefer calling `IViewOpenerService` as `IViewService` as it is dealing with views and the current `IViewOpenerService` as `IViewDescriptorService` as it is dealing with view descriptors.
